src/pool/srv_pool.c

```diff
@@ -976,112 +983,115 @@ pool_svc_free_cb(struct ds_rsvc *rsvc)
 	D_FREE(svc);
 }
 
+/*
+ * Initialize and update svc->ps_pool with map_buf and map_version. This
+ * ensures that svc->ps_pool matches the latest pool map.
+ */
 static int
-pool_svc_step_up_cb(struct ds_rsvc *rsvc)
+init_svc_pool(struct pool_svc *svc, struct pool_buf *map_buf,
+	      uint32_t map_version)
 {
-	struct pool_svc		       *svc = pool_svc_obj(rsvc);
-	struct rdb_tx			tx;
-	struct ds_pool		       *pool;
-	d_rank_list_t		       *replicas = NULL;
-	struct pool_map		       *map = NULL;
-	uint32_t			map_version;
 	struct ds_pool_create_arg	arg;
-	d_rank_t			rank;
+	struct ds_pool		       *pool;
 	int				rc;
 
-	/* Read the pool map into map and map_version. */
+	arg.pca_map_version = map_version;
+	arg.pca_need_group = 1;
+	rc = ds_pool_lookup_create(svc->ps_uuid, &arg, &pool);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to get ds_pool: %d\n",
+			DP_UUID(svc->ps_uuid), rc);
+		return rc;
+	}
+	rc = ds_pool_tgt_map_update(pool, map_buf, map_version);
+	if (rc != 0) {
+		ds_pool_put(pool);
+		return rc;
+	}
+	D_ASSERT(svc->ps_pool == NULL);
+	svc->ps_pool = pool;
+	return 0;
+}
+
+/* Finalize svc->ps_pool. */
+static void
+fini_svc_pool(struct pool_svc *svc)
+{
+	D_ASSERT(svc->ps_pool != NULL);
+	ds_pool_put(svc->ps_pool);
+	svc->ps_pool = NULL;
+}
+
+static int
+pool_svc_step_up_cb(struct ds_rsvc *rsvc)
+{
+	struct pool_svc	       *svc = pool_svc_obj(rsvc);
+	struct rdb_tx		tx;
+	d_rank_list_t	       *replicas = NULL;
+	struct pool_buf	       *map_buf = NULL;
+	uint32_t		map_version;
+	bool			cont_svc_up = false;
+	d_rank_t		rank;
+	int			rc;
+
+	/* Read the pool map into map_buf and map_version. */
 	rc = rdb_tx_begin(rsvc->s_db, rsvc->s_term, &tx);
 	if (rc != 0)
 		goto out;
 	ABT_rwlock_rdlock(svc->ps_lock);
-	rc = read_map(&tx, &svc->ps_root, &map);
-	if (rc == 0)
-		rc = rdb_get_ranks(rsvc->s_db, &replicas);
-	ABT_rwlock_unlock(svc->ps_lock);
-	rdb_tx_end(&tx);
+	rc = read_map_buf(&tx, &svc->ps_root, &map_buf, &map_version);
 	if (rc != 0) {
 		if (rc == -DER_NONEXIST) {
-			D_DEBUG(DF_DSMS, DF_UUID": new db\n",
+			D_DEBUG(DB_MD, DF_UUID": new db\n",
 				DP_UUID(svc->ps_uuid));
-			rc = DER_UNINIT;
+			rc = +DER_UNINIT;
```

Comment:
What is the purpose of the + here?  Should it be - or is it some sort of terse documentation that you intended to make it positive?

Reply:
The "+" is a reminder that this is a special return value, not that the author forget a "-". (See ds_rsvc_class.sc_step_up comments.)

Reply:
maybe would be useful to have a comment here (though I guess maybe that defeats the purpose).